implementations/rust/ockam/ockam_entity/src/imp.rs

```diff
@@ -0,0 +1,260 @@
+        let event =
+            ProfileChangeHistory::find_last_key_event(self.change_events(), key_attributes)?
+                .clone();
+        Self::get_secret_key_from_event(key_attributes, &event, &mut self.vault)
+    }
+
+    fn get_public_key(&self, key_attributes: &KeyAttributes) -> ockam_core::Result<PublicKey> {
+        self.change_history.get_public_key(key_attributes)
+    }
+    fn get_root_secret(&mut self) -> ockam_core::Result<Secret> {
+        let public_key =
+            ProfileChangeHistory::get_current_profile_update_public_key(self.change_events())?;
+
+        let key_id = self.vault.compute_key_id_for_public_key(&public_key)?;
+        self.vault.get_secret_by_key_id(&key_id)
+    }
+}
+
+impl<V: ProfileVault> ProfileContacts for ProfileImpl<V> {
+    fn contacts(&self) -> &ContactsDb {
+        &self.contacts
+    }
+
+    fn to_contact(&self) -> Contact {
+        Contact::new(
+            self.identifier.clone(),
+            self.change_history.as_ref().to_vec(),
+        )
+    }
+
+    fn serialize_to_contact(&self) -> ockam_core::Result<Vec<u8>> {
+        let contact = self.to_contact();
+
+        Profile::serialize_contact(&contact)
+    }
+
+    fn get_contact(&self, id: &ProfileIdentifier) -> Option<&Contact> {
+        self.contacts.get(id)
+    }
+
+    fn verify_contact(&mut self, contact: &Contact) -> ockam_core::Result<()> {
+        contact.verify(&mut self.vault)
+    }
+
+    fn verify_and_add_contact(&mut self, contact: Contact) -> ockam_core::Result<()> {
+        self.verify_contact(&contact)?;
+
+        let _ = self.contacts.insert(contact.identifier().clone(), contact);
+
+        Ok(())
+    }
+
+    fn verify_and_update_contact(
+        &mut self,
+        profile_id: &ProfileIdentifier,
+        change_events: Vec<ProfileChangeEvent>,
+    ) -> ockam_core::Result<()> {
+        let contact = self
+            .contacts
+            .get_mut(profile_id)
+            .ok_or(EntityError::ContactNotFound)?;
+
+        contact.verify_and_update(change_events, &mut self.vault)
+    }
+}
+
+impl<V: ProfileVault> ProfileAuth for ProfileImpl<V> {
+    /// Generate Proof of possession of [`Profile`].
+    /// channel_state should be tied to channel's cryptographical material (e.g. h value for Noise XX)
+    fn generate_authentication_proof(
+        &mut self,
+        channel_state: &[u8],
+    ) -> ockam_core::Result<Vec<u8>> {
+        let root_secret = self.get_root_secret()?;
+
+        Authentication::generate_proof(channel_state, &root_secret, &mut self.vault)
+    }
+
+    /// Verify Proof of possession of [`Profile`] with given [`ProfileIdentifier`].
+    /// channel_state should be tied to channel's cryptographical material (e.g. h value for Noise XX)
+    fn verify_authentication_proof(
+        &mut self,
+        channel_state: &[u8],
+        responder_contact_id: &ProfileIdentifier,
+        proof: &[u8],
```

Comment:
I know that this was in the original implementation too, but can we also add types here for the  channel state and proof parameters?